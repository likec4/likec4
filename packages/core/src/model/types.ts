import type { Tagged } from 'type-fest'
import type {
  AnyAux,
  ComputedLikeC4ModelData,
  ComputedView,
  DiagramView,
  LayoutedLikeC4ModelData,
  ParsedLikeC4ModelData,
  scalar,
  ViewType,
  ViewWithType,
} from '../types'
import type * as aux from '../types/aux'
import type { LikeC4ViewModel } from './view/LikeC4ViewModel'

export type IncomingFilter = 'all' | 'direct' | 'to-descendants'
export type OutgoingFilter = 'all' | 'direct' | 'from-descendants'

type AnyScalar<V extends string> =
  | Tagged<V, 'Fqn'>
  | Tagged<V, 'DeploymentFqn'>
  | Tagged<V, 'ViewId'>
  | Tagged<V, 'NodeId'>
  | Tagged<V, 'EdgeId'>
  | Tagged<V, 'RelationId'>

/**
 * Utility function to extract `id` from the given element.
 */
export function getId<Id extends string, Scalar extends AnyScalar<Id>>(element: Id | { id: Scalar }): Scalar {
  return typeof element === 'string' ? element as unknown as Scalar : element.id
}

export type ElementOrFqn<A extends AnyAux> = aux.ElementId<A> | { id: aux.Fqn<A> }

export type DeploymentOrFqn<A extends AnyAux> = aux.DeploymentId<A> | { id: aux.DeploymentFqn<A> }

export type ViewOrId<A extends AnyAux> = aux.ViewId<A> | { id: aux.StrictViewId<A> }

export type NodeOrId = string | { id: scalar.NodeId }
export type EdgeOrId = string | { id: scalar.EdgeId }
export type RelationOrId = string | { id: scalar.RelationId }

export type $View<A extends AnyAux> = {
  parsed: never
  computed: ComputedView<A>
  layouted: DiagramView<A>
}[aux.Stage<A>]

export type $ViewWithType<A extends AnyAux, V extends ViewType> = ViewWithType<$View<A>, V>

export type $ViewModel<A extends AnyAux> = {
  parsed: never
  computed: LikeC4ViewModel<A>
  layouted: LikeC4ViewModel<A>
}[aux.Stage<A>]

// export type $ModelData<A> = A extends infer T extends Any ? {
export type $ModelData<A extends AnyAux> = {
  parsed: ParsedLikeC4ModelData<A>
  computed: ComputedLikeC4ModelData<A>
  layouted: LayoutedLikeC4ModelData<A>
}[aux.Stage<A>]
