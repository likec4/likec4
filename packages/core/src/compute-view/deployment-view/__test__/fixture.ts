import { map, prop } from 'remeda'
import type { DeploymentRulesBuilderOp } from '../../../builder'
import { Builder } from '../../../builder'
import { LikeC4Model } from '../../../model'
import type { DeploymentView } from '../../../types'
import { withReadableEdges } from '../../utils/with-readable-edges'
import { computeDeploymentView } from '../compute'

const {
  builder: b,
  model: {
    model,
    person,
    system,
    component,
    webapp,
    mobile,
    ...$m
  },
  deployment: {
    deployment,
    env,
    node,
    zone,
    instanceOf,
    ...$d
  },
  views: {
    views,
    deploymentView,
    $rules,
    $include,
    $exclude,
  },
} = Builder.forSpecification({
  elements: {
    person: {
      style: {
        shape: 'person',
      },
    },
    system: {},
    component: {},
    webapp: {
      style: {
        shape: 'browser',
      },
    },
    mobile: {
      style: {
        shape: 'mobile',
      },
    },
  },
  deployments: {
    env: {},
    zone: {},
    node: {},
  },
  tags: ['old', 'next', 'temp'],
  relationships: { 'https': {} },
})

export const builder = b
  .with(
    model(
      person('customer', {
        title: 'Happy Customer',
      }),
      system('cloud').with(
        component('frontend').with(
          webapp('dashboard'),
          mobile('mobile'),
        ),
        component('auth'),
        component('backend').with(
          component('api'),
        ),
        component('media', {
          shape: 'storage',
        }),
      ),
      system('aws').with(
        component('rds', {
          shape: 'storage',
        }),
        component('s3', {
          shape: 'storage',
        }),
      ),
      system('email'),
    ),
    model(
      $m.rel('customer', 'cloud', 'uses services'),
      $m.rel('customer', 'cloud.frontend.mobile', 'opens mobile app'),
      $m.rel('customer', 'cloud.frontend.dashboard', {
        title: 'opens in browser',
        tags: ['old'],
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.auth', {
        title: 'authenticates',
        color: 'green',
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.backend.api', {
        title: 'fetches data',
        tags: ['old'],
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.media', {
        title: 'fetches media',
        tags: ['temp'],
      }),
      $m.rel('cloud.frontend.mobile', 'cloud.auth', {
        title: 'authenticates',
        color: 'amber',
      }),
      $m.rel('cloud.frontend.mobile', 'cloud.backend.api', 'fetches data'),
      $m.rel('cloud.frontend.mobile', 'cloud.media', 'fetches media'),
      $m.rel('cloud.backend.api', 'cloud.auth', 'authorizes'),
      $m.rel('cloud.backend.api', 'cloud.media', 'uploads media'),
      $m.rel('cloud.backend.api', 'aws.rds', 'reads/writes'),
      $m.rel('cloud.backend.api', 'email', 'sends emails'),
      $m.rel('cloud.media', 'aws.s3', 'uploads'),
      $m.rel('email', 'customer', 'sends emails'),
    ),
    deployment(
      node('customer').with(
        instanceOf('instance', 'customer'),
      ),
      env('prod').with(
        zone('eu').with(
          zone('zone1').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          zone('zone2').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          instanceOf('auth', 'cloud.auth'),
          instanceOf('media', 'cloud.media'),
          instanceOf('db', 'aws.rds'),
        ),
        zone('us').with(
          zone('zone1').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          instanceOf('db', 'aws.rds'),
        ),
      ),
      env('dev').with(
        node('devCustomer').with(
          instanceOf('instance', 'customer'),
        ),
        node('devCloud').with(
          instanceOf('instance', 'cloud'),
        ),
      ),
      env('acc').with(
        node('testCustomer').with(
          instanceOf('instance', 'customer'),
        ),
        zone('eu').with(
          instanceOf('api', 'cloud.backend.api'),
          instanceOf('ui', 'cloud.frontend.dashboard'),
          instanceOf('auth', 'cloud.auth'),
          instanceOf('db', 'aws.rds'),
        ),
      ),
      env('global').with(
        instanceOf('email', 'email'),
      ),
      $d.rel('prod.eu.db', 'prod.us.db', {
        title: 'replicates',
        color: 'green',
      }),
      $d.rel('prod.us.db', 'prod.eu.db', {
        title: 'replicates',
        color: 'amber',
      }),
    ),
  )
  .clone()

export type Types = typeof builder['Types']

export { $exclude, $include }

export function computeView(...rules: DeploymentRulesBuilderOp<Types>[]) {
  const modelsource = builder.clone()
    .with(
      views(
        deploymentView('index', $rules(...rules)),
      ),
    ).build()

  const model = LikeC4Model.create({
    ...modelsource,
    views: {},
  })

  const view = withReadableEdges(computeDeploymentView(model, modelsource.views.index as DeploymentView))

  return Object.assign(view, {
    nodeIds: map(view.nodes, prop('id')) as string[],
    edgeIds: map(view.edges, prop('id')) as string[],
  })
}

export function computeNodesAndEdges(...rules: DeploymentRulesBuilderOp<Types>[]) {
  const { nodeIds, edgeIds } = computeView(...rules)
  return {
    // Starts with capital letter to be first in snapshot
    Nodes: nodeIds,
    edges: edgeIds,
  }
}

export function createModel() {
  const modelsource = builder.build()

  return LikeC4Model.create({
    ...modelsource,
    views: {},
  })
}
