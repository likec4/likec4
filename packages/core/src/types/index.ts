export * from './_common'

export type * as aux from './aux'

export type {
  Any,
  AnyAux,
  Aux,
  SpecAux,
  Unknown,
  UnknownComputed,
  UnknownLayouted,
  UnknownParsed,
} from './aux'

export * from './const'
export * from './expression'
export * from './expression-model'
export * from './fqnRef'
export * from './geometry'
export * from './global'
export * from './model-data'
export * from './model-deployment'
export * from './model-dump'
export * from './model-logical'
export * from './model-spec'
export * from './operators'
export * from './scalar'
export * from './styles'
export * from './view'
export * from './view-changes'
export * from './view-common'
export * from './view-computed'
export * from './view-layouted'
export * from './view-parsed.deployment'
export * from './view-parsed.dynamic'
export * from './view-parsed.element'

export type * as scalar from './scalar'
