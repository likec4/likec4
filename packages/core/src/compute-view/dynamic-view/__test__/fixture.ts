import { partition } from 'remeda'
import type {
  DynamicViewIncludeRule,
  DynamicViewRule,
  DynamicViewStep,
  ElementViewPredicate,
  Fqn,
  ParsedDynamicView as DynamicView,
  ViewId,
} from '../../../types'
import { type $Aux, type FakeElementIds, fakeModel } from '../../element-view/__test__/fixture'
import { computeDynamicView } from '../compute'

const emptyView = {
  _type: 'dynamic' as const,
  id: 'index' as ViewId,
  title: null,
  description: null,
  tags: null,
  links: null,
  rules: [],
}

type StepExpr = `${FakeElementIds} ${'->' | '<-'} ${FakeElementIds}`
type StepProps = Omit<DynamicViewStep, 'source' | 'target' | 'isBackward'>

export function $step(expr: StepExpr, props?: string | Partial<StepProps>): DynamicViewStep {
  const title = typeof props === 'string' ? props : props?.title
  if (expr.includes(' -> ')) {
    const [source, target] = expr.split(' -> ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      ...(typeof props === 'object' ? props : {}),
      title: title ?? null,
    }
  }
  if (expr.includes(' <- ')) {
    const [target, source] = expr.split(' <- ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      ...(typeof props === 'object' ? props : {}),
      title: title ?? null,
      isBackward: true,
    }
  }
  throw new Error(`Invalid step expression: ${expr}`)
}

export function compute(
  stepsAndRules: (DynamicViewStep<$Aux> | ElementViewPredicate<$Aux> | DynamicViewIncludeRule<$Aux>)[],
) {
  const [steps, rules] = partition(stepsAndRules, (s): s is DynamicViewStep => 'source' in s)
  let view = computeDynamicView(
    fakeModel,
    {
      ...emptyView,
      steps,
      rules: rules as DynamicViewRule[],
    } as DynamicView,
  )
  return Object.assign(view, {
    nodeIds: view.nodes.map((node) => node.id) as string[],
    edgeIds: view.edges.map((edge) => edge.id) as string[],
  })
}
