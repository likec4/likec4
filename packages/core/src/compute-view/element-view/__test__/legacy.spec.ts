import { describe, it } from 'vitest'
import { type IconUrl } from '../../../types'
import { $exclude, $expr, $include, computeView } from './fixture'

describe('compute-element-view', () => {
  it('should be empty if no root and no rules', ({ expect }) => {
    const { nodes, edges } = computeView([])
    expect.soft(nodes).toEqual([])
    expect(edges).toEqual([])
  })

  it('should show only root if no rules', ({ expect }) => {
    const { nodeIds, edges } = computeView('cloud', [])
    expect(nodeIds).toEqual(['cloud'])
    expect(edges).toEqual([])
  })

  it('should return landscape view on top `include *`', ({ expect }) => {
    const { nodes, nodeIds, edgeIds } = computeView([$include('*')])
    expect.soft(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'email',
      'amazon',
    ])
    expect(edgeIds).toEqual([
      'customer:cloud',
      'support:cloud',
      'cloud:email',
      'cloud:amazon',
      'email:cloud',
    ])
    const [customer, support, cloud, , amazon] = nodes
    expect.soft(amazon).toMatchObject({
      id: 'amazon',
      outEdges: [],
      inEdges: ['cloud:amazon'],
    })
    expect.soft(cloud).toMatchObject({
      id: 'cloud',
      inEdges: [
        'customer:cloud',
        'support:cloud',
        'email:cloud',
      ],
      outEdges: [
        'cloud:email',
        'cloud:amazon',
      ],
    })
    expect.soft(customer).toMatchObject({
      id: 'customer',
      outEdges: ['customer:cloud'],
      inEdges: [],
    })
    expect.soft(support).toMatchObject({
      id: 'support',
      outEdges: ['support:cloud'],
      inEdges: [],
    })
  })

  // TODO investigate why this test fails
  it.fails('should return nodes in the same order as was in view', ({ expect }) => {
    const { nodeIds, edgeIds } = computeView([$include('email'), $include('*')])
    expect(nodeIds).toEqual([
      'email',
      'customer',
      'support',
      'cloud',
      'amazon',
    ])
    expect(edgeIds).toEqual(['customer:cloud', 'support:cloud', 'cloud:amazon'])
  })

  it('view of cloud', ({ expect }) => {
    const { nodeIds, edgeIds, ...view } = computeView('cloud', [$include('*')])
    expect.soft(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'cloud.frontend',
      'cloud.backend',
      'email',
      'amazon',
    ])

    expect(edgeIds).to.have.same.members([
      'customer:cloud.frontend',
      'support:cloud.frontend',
      'cloud.frontend:cloud.backend',
      'cloud.backend:amazon',
      'cloud.backend:email',
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of cloud.backend', ({ expect }) => {
    const { nodeIds, edgeIds } = computeView('cloud.backend', [
      $include('*'),
      $include('customer'),
    ])

    expect.soft(nodeIds).toEqual([
      'customer',
      'cloud.frontend',
      'cloud.backend',
      'cloud.backend.graphql',
      'cloud.backend.storage',
      'amazon',
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.backend.graphql:cloud.backend.storage',
      'cloud.backend.storage:amazon',
      'cloud.frontend:cloud.backend.graphql',
      'customer:cloud.frontend',
    ])
  })

  it('view of cloud.frontend', ({ expect }) => {
    const { nodeIds, edgeIds, ...view } = computeView('cloud.frontend', [$include('*')])
    expect.soft(nodeIds).toEqual([
      'customer',
      'support',
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.frontend.supportPanel',
      'cloud.backend',
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend.supportPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend',
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.supportPanel',
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of amazon', ({ expect }) => {
    const { nodeIds, edgeIds } = computeView('amazon', [$include('*')])

    expect(nodeIds).toEqual(['cloud', 'amazon', 'amazon.s3'])
    expect(edgeIds).toEqual(['cloud:amazon.s3'])
  })

  it('view of amazon including cloud.* ->', ({ expect }) => {
    const { edgeIds, nodeIds } = computeView('amazon', [$include('*'), $include('cloud.* ->')])

    expect(nodeIds).toEqual([
      'cloud',
      'cloud.backend',
      'amazon',
      'amazon.s3',
    ])

    expect(edgeIds).to.have.same.members(['cloud.backend:amazon.s3'])
  })

  it('view of cloud.frontend (and include parent cloud)', ({ expect }) => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [$include('*'), $include('cloud')])

    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.frontend.supportPanel',
      'cloud.backend',
    ])

    expect(edgeIds).to.have.same.members([
      'support:cloud.frontend.supportPanel',
      'customer:cloud.frontend.dashboard',
      'cloud.frontend.supportPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend',
    ])
  })

  it('view of cloud.frontend (and include cloud.backend.*)', ({ expect }) => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [
      $include('*'),
      $include('cloud'),
      $include('cloud.backend.*'),
    ])

    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.frontend.supportPanel',
      'cloud.backend',
      'cloud.backend.graphql',
      'cloud.backend.storage',
    ])

    expect(edgeIds).toEqual([
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.supportPanel',
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.frontend.supportPanel:cloud.backend.graphql',
      'cloud.backend.graphql:cloud.backend.storage',
    ])
  })

  it('view of cloud.frontend (and include cloud.frontend -> cloud.backend.*)', ({ expect }) => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [
      $include('*'),
      $include('cloud'),
      $include('cloud.frontend -> cloud.backend.*'),
    ])

    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.frontend.supportPanel',
      'cloud.backend',
      'cloud.backend.graphql',
    ])

    expect(edgeIds).to.have.same.members([
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.supportPanel',
      'cloud.frontend.supportPanel:cloud.backend.graphql',
      'cloud.frontend.dashboard:cloud.backend.graphql',
    ])
  })

  it('view of cloud (exclude cloud, amazon)', ({ expect }) => {
    const { edgeIds, nodeIds } = computeView('cloud', [
      $include('*'),
      $exclude('cloud'),
      $exclude('email'),
      $exclude('amazon'),
    ])

    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud.frontend',
      'cloud.backend',
    ])

    expect(edgeIds).toEqual([
      'customer:cloud.frontend',
      'support:cloud.frontend',
      'cloud.frontend:cloud.backend',
    ])
  })

  it('view with 3 levels', ({ expect }) => {
    const { edgeIds, nodeIds } = computeView('cloud', [
      $include('*'),
      $include('cloud.frontend.*'),
      $include('cloud.backend.*'),
      $include('-> amazon.*'),
      $exclude('amazon'),
      $exclude('email'),
      $exclude('cloud.frontend'),
    ])

    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'cloud.frontend.dashboard',
      'cloud.frontend.supportPanel',
      'cloud.backend',
      'cloud.backend.graphql',
      'cloud.backend.storage',
      'amazon.s3',
    ])

    expect(edgeIds).toEqual([
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.supportPanel',
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.frontend.supportPanel:cloud.backend.graphql',
      'cloud.backend.graphql:cloud.backend.storage',
      'cloud.backend.storage:amazon.s3',
    ])
  })

  it('view with 3 levels (with only relevant elements)', ({ expect }) => {
    const { edgeIds, nodeIds } = computeView('cloud', [
      $include('*'),
      $include('-> cloud.frontend.*'),
      $include('-> cloud.backend.*'),
    ])

    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.frontend.supportPanel',
      'cloud.backend',
      'cloud.backend.graphql',
      'email',
    ])

    expect(edgeIds).toEqual([
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.supportPanel',
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.frontend.supportPanel:cloud.backend.graphql',
      'cloud.backend:email',
    ])
  })

  it('index view with applied styles', ({ expect }) => {
    const { nodes } = computeView([
      $include('customer'),
      $include('amazon'),
      $include('cloud'),
      $include('cloud.frontend.dashboard'),
      // all elements
      // color: secondary
      {
        targets: [{ wildcard: true }],
        style: {
          color: 'secondary',
          icon: 'none',
          shape: 'storage',
        },
      },
      // cloud
      // color: muted
      {
        targets: [$expr('cloud')],
        style: {
          color: 'muted',
          icon: 'http://some-icon' as IconUrl,
        },
      },
      // cloud.*
      // shape: browser
      {
        targets: [$expr('cloud.*')],
        style: {
          shape: 'browser',
        },
      },
    ])

    const amazon = nodes.find(n => n.id === 'amazon')!
    const customer = nodes.find(n => n.id === 'customer')!
    const cloud = nodes.find(n => n.id === 'cloud')!
    const frontend = nodes.find(n => n.id === 'cloud.frontend.dashboard')!

    expect(amazon).toMatchObject({
      color: 'secondary',
      shape: 'storage',
    })
    expect(amazon).not.toHaveProperty('icon')
    expect(customer).toMatchObject({
      color: 'secondary',
      shape: 'storage',
    })
    expect(customer).not.toHaveProperty('icon')

    expect(cloud).toMatchObject({
      color: 'muted',
      shape: 'storage',
      icon: 'http://some-icon',
    })
    expect(frontend).toMatchObject({
      id: 'cloud.frontend.dashboard',
      parent: 'cloud',
      color: 'secondary',
      shape: 'browser',
    })
    expect(frontend).not.toHaveProperty('icon')
  })

  it('should include by element kind', ({ expect }) => {
    const { nodeIds, edgeIds } = computeView({
      include: [
        {
          elementKind: 'system',
          isEqual: true,
        },
      ],
    })

    expect(nodeIds).toEqual([
      'cloud',
      'email',
      'amazon',
    ])
    expect(edgeIds).toEqual([
      'cloud:email',
      'cloud:amazon',
      'email:cloud',
    ])
  })

  it('should include by element tag', ({ expect }) => {
    const { nodeIds, edgeIds } = computeView({
      include: [
        {
          elementTag: 'old',
          isEqual: true,
        },
      ],
    })

    expect(nodeIds).toEqual(['cloud', 'cloud.backend.storage', 'cloud.frontend.supportPanel'])

    expect(edgeIds).toEqual([])
  })

  it('should exclude by element kind', ({ expect }) => {
    const { nodeIds, edgeIds } = computeView('cloud.frontend', [
      $include('*'),
      {
        exclude: [
          {
            elementKind: 'actor',
            isEqual: true,
          },
        ],
      },
    ])

    expect(nodeIds).toEqual([
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.frontend.supportPanel',
      'cloud.backend',
    ])

    expect(edgeIds).toEqual([
      'cloud.frontend.dashboard:cloud.backend',
      'cloud.frontend.supportPanel:cloud.backend',
    ])
  })

  it('should exclude by element tag', ({ expect }) => {
    const { nodeIds, edgeIds } = computeView('cloud.backend', [
      $include('*'),
      {
        exclude: [
          {
            elementTag: 'old',
            isEqual: true,
          },
        ],
      },
    ])

    expect(nodeIds).toEqual([
      'cloud.frontend',
      'cloud.backend',
      'cloud.backend.graphql',
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend:cloud.backend.graphql',
    ])
  })
})
