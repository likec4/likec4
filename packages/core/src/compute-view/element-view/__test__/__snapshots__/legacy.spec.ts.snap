// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compute-element-view > view of cloud 1`] = `
{
  "_stage": "computed",
  "_type": "element",
  "autoLayout": {
    "direction": "TB",
  },
  "description": null,
  "edges": [
    {
      "id": "customer:cloud.frontend",
      "label": "opens in browser",
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend",
    },
    {
      "id": "support:cloud.frontend",
      "label": "manages",
      "parent": null,
      "relations": [
        "support:cloud.frontend.supportPanel",
      ],
      "source": "support",
      "target": "cloud.frontend",
    },
    {
      "id": "cloud.frontend:cloud.backend",
      "kind": "graphlql",
      "label": "requests",
      "parent": "cloud",
      "relations": [
        "cloud.frontend:cloud.backend",
        "cloud.frontend.dashboard:cloud.backend.graphql",
        "cloud.frontend.supportPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend",
      "tags": [
        "next",
        "old",
      ],
      "tail": "odiamond",
      "target": "cloud.backend",
    },
    {
      "id": "cloud.backend:email",
      "label": "schedule",
      "parent": null,
      "relations": [
        "cloud.backend:email",
      ],
      "source": "cloud.backend",
      "tags": [
        "communication",
      ],
      "target": "email",
    },
    {
      "id": "cloud.backend:amazon",
      "label": "uploads",
      "parent": null,
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud.backend",
      "tags": [
        "aws",
        "storage",
        "legacy",
      ],
      "target": "amazon",
    },
  ],
  "hash": "6df052156ae59d91622677537c2d657e4946d803",
  "id": "index",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "modelRef": "customer",
      "outEdges": [
        "customer:cloud.frontend",
      ],
      "parent": null,
      "shape": "person",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "customer",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "modelRef": "support",
      "outEdges": [
        "support:cloud.frontend",
      ],
      "parent": null,
      "shape": "person",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "support",
    },
    {
      "children": [
        "cloud.frontend",
        "cloud.backend",
      ],
      "color": "primary",
      "depth": 1,
      "description": null,
      "id": "cloud",
      "inEdges": [
        "customer:cloud.frontend",
        "support:cloud.frontend",
      ],
      "kind": "system",
      "level": 0,
      "links": null,
      "modelRef": "cloud",
      "outEdges": [
        "cloud.backend:email",
        "cloud.backend:amazon",
      ],
      "parent": null,
      "shape": "rectangle",
      "style": {},
      "tags": [
        "next",
        "old",
      ],
      "technology": null,
      "title": "cloud",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend",
      "inEdges": [
        "customer:cloud.frontend",
        "support:cloud.frontend",
      ],
      "kind": "container",
      "level": 1,
      "links": null,
      "modelRef": "cloud.frontend",
      "outEdges": [
        "cloud.frontend:cloud.backend",
      ],
      "parent": "cloud",
      "shape": "browser",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "frontend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "inEdges": [
        "cloud.frontend:cloud.backend",
      ],
      "kind": "container",
      "level": 1,
      "links": null,
      "modelRef": "cloud.backend",
      "outEdges": [
        "cloud.backend:email",
        "cloud.backend:amazon",
      ],
      "parent": "cloud",
      "shape": "rectangle",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "backend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "email",
      "inEdges": [
        "cloud.backend:email",
      ],
      "kind": "system",
      "level": 0,
      "links": null,
      "modelRef": "email",
      "outEdges": [],
      "parent": null,
      "shape": "rectangle",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "email",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "icon": "tech:aws",
      "id": "amazon",
      "inEdges": [
        "cloud.backend:amazon",
      ],
      "kind": "system",
      "level": 0,
      "links": null,
      "modelRef": "amazon",
      "outEdges": [],
      "parent": null,
      "shape": "rectangle",
      "style": {},
      "tags": [
        "aws",
      ],
      "technology": null,
      "title": "amazon",
    },
  ],
  "tags": null,
  "title": null,
  "viewOf": "cloud",
}
`;

exports[`compute-element-view > view of cloud.frontend 1`] = `
{
  "_stage": "computed",
  "_type": "element",
  "autoLayout": {
    "direction": "TB",
  },
  "description": null,
  "edges": [
    {
      "id": "customer:cloud.frontend.dashboard",
      "label": "opens in browser",
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend.dashboard",
    },
    {
      "id": "support:cloud.frontend.supportPanel",
      "label": "manages",
      "parent": null,
      "relations": [
        "support:cloud.frontend.supportPanel",
      ],
      "source": "support",
      "target": "cloud.frontend.supportPanel",
    },
    {
      "id": "cloud.frontend.dashboard:cloud.backend",
      "kind": "graphlql",
      "label": "requests",
      "line": "solid",
      "parent": null,
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.dashboard",
      "tags": [
        "next",
      ],
      "target": "cloud.backend",
    },
    {
      "id": "cloud.frontend.supportPanel:cloud.backend",
      "kind": "graphlql",
      "label": "fetches",
      "line": "dashed",
      "parent": null,
      "relations": [
        "cloud.frontend.supportPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.supportPanel",
      "tags": [
        "old",
      ],
      "tail": "odiamond",
      "target": "cloud.backend",
    },
  ],
  "hash": "388ed63a4179b91d2953e5c4908c651e3e826eb0",
  "id": "index",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "modelRef": "customer",
      "outEdges": [
        "customer:cloud.frontend.dashboard",
      ],
      "parent": null,
      "shape": "person",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "customer",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "modelRef": "support",
      "outEdges": [
        "support:cloud.frontend.supportPanel",
      ],
      "parent": null,
      "shape": "person",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "support",
    },
    {
      "children": [
        "cloud.frontend.dashboard",
        "cloud.frontend.supportPanel",
      ],
      "color": "primary",
      "depth": 1,
      "description": null,
      "id": "cloud.frontend",
      "inEdges": [
        "customer:cloud.frontend.dashboard",
        "support:cloud.frontend.supportPanel",
      ],
      "kind": "container",
      "level": 0,
      "links": null,
      "modelRef": "cloud.frontend",
      "outEdges": [
        "cloud.frontend.dashboard:cloud.backend",
        "cloud.frontend.supportPanel:cloud.backend",
      ],
      "parent": null,
      "shape": "browser",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "frontend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "icon": "tech:react",
      "id": "cloud.frontend.dashboard",
      "inEdges": [
        "customer:cloud.frontend.dashboard",
      ],
      "kind": "component",
      "level": 1,
      "links": null,
      "modelRef": "cloud.frontend.dashboard",
      "outEdges": [
        "cloud.frontend.dashboard:cloud.backend",
      ],
      "parent": "cloud.frontend",
      "shape": "rectangle",
      "style": {},
      "tags": [
        "next",
      ],
      "technology": null,
      "title": "dashboard",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend.supportPanel",
      "inEdges": [
        "support:cloud.frontend.supportPanel",
      ],
      "kind": "component",
      "level": 1,
      "links": null,
      "modelRef": "cloud.frontend.supportPanel",
      "outEdges": [
        "cloud.frontend.supportPanel:cloud.backend",
      ],
      "parent": "cloud.frontend",
      "shape": "rectangle",
      "style": {},
      "tags": [
        "old",
      ],
      "technology": null,
      "title": "adminPanel",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "inEdges": [
        "cloud.frontend.dashboard:cloud.backend",
        "cloud.frontend.supportPanel:cloud.backend",
      ],
      "kind": "container",
      "level": 0,
      "links": null,
      "modelRef": "cloud.backend",
      "outEdges": [],
      "parent": null,
      "shape": "rectangle",
      "style": {},
      "tags": [],
      "technology": null,
      "title": "backend",
    },
  ],
  "tags": null,
  "title": null,
  "viewOf": "cloud.frontend",
}
`;
